meson.build

```diff
@@ -0,0 +1,1629 @@
+
+win32 = host_machine.system() == 'cygwin' or host_machine.system() == 'windows'
+mingw = win32 # assume true for now but this is probably wrong
+posix = false
+if not win32
+    posix = true
+endif
+
+mswin_flags = ['-D_WIN32_WINNT=0x0602', '-DUNICODE', '-DCOBJMACROS',
+               '-DINITGUID', '-U__STRICT_ANSI__']
+
+mswin_linkflags = ['-Wl,--major-os-version=6,--minor-os-version=0',
+                   '-Wl,--major-subsystem-version=6,--minor-subsystem-version=0']
+
+if host_machine.system() == 'windows'
+    flags += [mswin_flags, '-D__USE_MINGW_ANSI_STDIO=1']
+    link_flags += [mswin_linkflags, '-mwindows']
+endif
+
+if host_machine.system() == 'cygwin'
+    flags += [mswin_flags, '-mwin32']
+    link_flags += mswin_linkflags
+endif
+
+noexecstack = false
+if cc.has_link_argument('-Wl,-z,noexecstack')
+    link_flags += '-Wl,-z,noexecstack'
+    noexecstack = true
+endif
+
+if cc.has_link_argument('-Wl,--nxcompat,--no-seh,--dynamicbase')
+    link_flags += '-Wl,--nxcompat,--no-seh,--dynamicbase'
+    noexecstack = true
+endif
+
+if get_option('build-date').disabled()
+    flags += '-DNO_BUILD_TIMESTAMPS'
+endif
+
+if get_option('pthread-debug').enabled()
+    if cc.has_argument('-DMP_PTHREAD_DEBUG')
+        flags += '-DMP_PTHREAD_DEBUG'
+    endif
+endif
+
+libdl = {
+    'name': 'libdl',
+    'deps': cc.find_library('libdl', required: false),
+    'use' : false,
+}
+if libdl['deps'].found()
+    libdl_fragment = run_command('cat', 'waftools/fragments/libdl.c').stdout()
+    libdl += {'use': cc.compiles(libdl_fragment, dependencies: libdl['deps'], name: 'libdl check')}
+endif
+if libdl['use']
+    dependencies += libdl['deps']
+endif
+
+cplugins = {
+    'name': 'cplugins',
+    'use': libdl['use'] and not win32 and not get_option('cplugins').disabled()
+}
+if cplugins['use']
+    if cc.has_link_argument('-rdynamic')
+        link_flags += '-rdynamic'
+    endif
+endif
+
+add_project_arguments(flags, language: 'c')
+add_project_link_arguments(link_flags, language: 'c')
+
+
+# various file generators
+tools_directory = join_paths(meson.source_root(), 'TOOLS')
+file2string = join_paths(tools_directory, 'file2string.py')
+generated_dir = join_paths(meson.build_root(), 'generated')
+
+run_command('mkdir', '-p', join_paths(generated_dir, 'etc'))
+run_command('mkdir', '-p', join_paths(generated_dir, 'sub'))
+run_command('mkdir', '-p', join_paths(generated_dir, 'TOOLS', 'osxbundle', 'mpv.app', 'Contents', 'Resources'))
+run_command('mkdir', '-p', join_paths(generated_dir, 'player', 'lua'))
+run_command('mkdir', '-p', join_paths(generated_dir, 'player', 'javascript'))
+
+osxbundle = ['icon.icns']
+foreach file: osxbundle
+    input = join_paths(meson.source_root(), 'TOOLS', 'osxbundle', 'mpv.app', 'Contents', 'Resources', file)
+    output = join_paths(generated_dir, 'TOOLS', 'osxbundle', 'mpv.app', 'Contents', 'Resources', file + '.inc')
+    command = 'python '+file2string+' '+input+' > '+output
+    run_command('sh', '-c', command)
```

Comment:
IMO you should not write this as an sh -c thing... just do `run_command(find_program('python'), input, output)` and python will be run with two arguments and not depend on sh for Windows builds. It is also slightly more efficient, and produces slightly less convoluted/escaped log messages about the command being run.

You may need to rewrite file2string to support accepting the outfile as sys.argv[2], but that can trivially be done in a backwards compatible way by accepting both.

OTOH you could also probably only do this if `host_machine.system() == 'darwin'`.

...

Why is this even a run_command? Why not make it a custom_target? The latter will be re-run every time the input file changes, rather than only when running `meson setup`.